### ISSUE TITLE
attemp to kill datalad process results in breeding git <defunct> processes

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#2041]
#### What is the problem?
As reported in #2036 aggregate-metadata got stuck, decided to Ctrl-C it but it keeps "resisting" -- doesn't quit:

```
Exception ignored in: <bound method AnnexRepo.__del__ of <AnnexRepo path=/etc (<class 'datalad.support.annexrepo.AnnexRepo'>)>>
Traceback (most recent call last):
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 323, in __del__



^C^CException ignored in: <bound method GitRepo.__del__ of <GitRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.gitrepo.GitRepo'>)>>
Traceback (most recent call last):
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 732, in __del__
    if hasattr(self, 'repo') and exists(self.path) \
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 618, in repo
    self._repo = self.cmd_call_wrapper(Repo, self.path)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 134, in __call__
    return self.call(cmd, *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 492, in call
    return f(*args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 387, in Repo
    return gitpy.Repo(*args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/repo/base.py", line 165, in __init__
    self._bare = self.config_reader("repository").getboolean('core', 'bare')
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/config.py", line 78, in assure_data_present
    self.read()
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/config.py", line 404, in read
    with open(file_path, 'rb') as fp:
  File "/home/yoh/proj/datalad/datalad-master/datalad/auto.py", line 165, in _proxy_open
    *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/auto.py", line 149, in _proxy_open_name_mode
    self._dataset_auto_get(file)
  File "/home/yoh/proj/datalad/datalad-master/datalad/auto.py", line 205, in _dataset_auto_get
    annex = get_repo_instance(filedir)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmdline/helpers.py", line 229, in get_repo_instance
    return AnnexRepo(dir_, create=False)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/repo.py", line 146, in __call__
    instance = type.__call__(cls, *new_args, **new_kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 190, in __init__
    for r in self.get_remotes():
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 933, in get_remotes
    remotes = super(AnnexRepo, self).get_remotes(with_urls_only=with_urls_only)
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 1248, in get_remotes
    self.config.reload()
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 773, in config
    self._cfg = ConfigManager(dataset=self, dataset_only=False)
  File "/home/yoh/proj/datalad/datalad-master/datalad/config.py", line 198, in __init__
    self.reload(force=True)
  File "/home/yoh/proj/datalad/datalad-master/datalad/config.py", line 252, in reload
    stdout, stderr = self._run(run_args, log_stderr=True)
  File "/home/yoh/proj/datalad/datalad-master/datalad/config.py", line 520, in _run
    out = self._runner.run(['git', 'config'] + args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 588, in run
    cmd, env=self.get_git_environ_adjusted(env), *args, **kwargs)
  File "/home/yoh/proj/datalad/datalad-master/datalad/cmd.py", line 412, in run
    stdin=stdin)
  File "/usr/lib/python3.5/subprocess.py", line 676, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1244, in _execute_child
    part = os.read(errpipe_read, 50000)
KeyboardInterrupt:
^C^CException ignored in: <bound method AnnexRepo.__del__ of <AnnexRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.annexrepo.AnnexRepo'>)>>
Traceback (most recent call last):
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/annexrepo.py", line 323, in __del__
    super(AnnexRepo, self).__del__()
  File "/home/yoh/proj/datalad/datalad-master/datalad/support/gitrepo.py", line 739, in __del__
    self.repo.index.write()
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/index/base.py", line 216, in write
    self._serialize(stream, ignore_extension_data)
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/index/base.py", line 181, in _serialize
    write_cache(entries, stream, extension_data)
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/index/fun.py", line 146, in write_cache
    write(path)
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/git/util.py", line 708, in write
    return self.f.write(data)
  File "/home/yoh/proj/datalad/datalad-master/venvs/dev3/lib/python3.5/site-packages/gitdb/stream.py", line 695, in write
    os.write(self._fd, data)
KeyboardInterrupt:
^C^C^C^CException ignored in: <bound method AnnexRepo.__del__ of <AnnexRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.annexrepo.AnnexRepo'>)>>
Traceback (most recent call last):
  File "/home/yoh/proj/datalad/datalad-master/datalad/su....
```
checking list of processes showed that there is a growing number of git processes started by this process:
```shell
$> ps --ppid 1323230 -p 1323230 c -f
UID          PID    PPID  C STIME TTY      STAT   TIME CMD
yoh      1323230 2235280 44 Dec15 pts/10   R+   785:39 datalad
yoh      3232720 1323230  0 21:56 pts/10   Z+     0:00 git <defunct>
yoh      3233169 1323230  0 21:56 pts/10   Z+     0:00 git <defunct>
yoh      3233238 1323230  0 21:56 pts/10   Z+     0:00 git <defunct>
yoh      3239231 1323230  0 21:58 pts/10   Z+     0:00 git <defunct>
yoh      3239257 1323230  0 21:58 pts/10   Z+     0:00 git <defunct>
yoh      3239329 1323230  0 21:58 pts/10   Z+     0:00 git <defunct>
yoh      3239377 1323230  0 21:58 pts/10   Z+     0:00 git <defunct>
yoh      3239478 1323230  0 21:58 pts/10   Z+     0:00 git <defunct>
yoh      3265998 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3266012 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3266136 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3266739 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3266850 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3266878 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267003 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267063 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267273 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267297 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267420 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267479 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267870 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3267944 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>
yoh      3268112 1323230  0 22:05 pts/10   Z+     0:00 git <defunct>

```

Sure thing it must be specific to the ghosts of smaug

**Comment by yarikoptic** [comment]
ha -- with python2 it looks a "bit" different and may be revealing (also can't kill with Ctrl-C)
```shell
(git)smaug:/tmp/crawl[master]git
$> datalad search Haxby type:dataset
[INFO   ] Rebuilding search index
[INFO   ] Scanning for metadata keys
Exception IOError: IOError("Lock at '/home/yoh/proj/datalad/datalad-master/.git/index.lock' could not be obtained",) in <bound method AnnexRepo.__del__ of <AnnexRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.annexrepo.AnnexRepo'>)>> ignored
Exception IOError: IOError("Lock at '/home/yoh/proj/datalad/datalad-master/.git/index.lock' could not be obtained",) in <bound method GitRepo.__del__ of <GitRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.gitrepo.GitRepo'>)>> ignored
...
```
and the question (if you look at the path) is -- why the ... it is trying to deal with the datalad source repo (from which datalad was installed for development into the virtual environment)

**Comment by yarikoptic** [comment]
ha -- probably it is the AutomagicIO's fault. but while looking at it  I see lots of repetitive dance with the config:

```shell
[INFO   ] Rebuilding search index
[INFO   ] Scanning for metadata keys
[DEBUG  ] Activating DataLad's AutoMagicIO
[Level 5] Parsed ri /tmp/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/tmp/crawl', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 9] Running: ['git', 'annex', 'version', '--raw']
[Level 8] Finished running ['git', 'annex', 'version', '--raw'] with status 0
[Level 9] Running: ['git', 'version']
[Level 8] Finished running ['git', 'version'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Parsed ri /mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 5] Parsed ri /tmp/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/tmp/crawl', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 9] Running: ['git', 'version']
[Level 8] Finished running ['git', 'version'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Parsed ri /mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 8] Using existing Git repository at /tmp/crawl
[Level 5] Parsed ri /tmp/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/tmp/crawl', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 9] Running: ['git', 'version']
[Level 8] Finished running ['git', 'version'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin', '--file', '/tmp/crawl/.datalad/config'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Parsed ri /mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/mnt/btrfs/datasets-openfmri-crawl-20171120/datalad/crawl', 'password': '', 'port': ''}
....
[Level 8] Finished running ['git', 'version'] with status 0
[Level 9] Running: ['git', 'config', '-z', '-l', '--show-origin']
[Level 8] Finished running ['git', 'config', '-z', '-l', '--show-origin'] with status 0
[Level 5] Parsed ri git://github.com/yarikoptic/datalad into fields {'username': '', 'scheme': 'git', 'fragment': '', 'hostname': 'github.com', 'query': '', 'path': '/yarikoptic/datalad', 'password': '', 'port': ''}
[Level 5] Detected url ri
[Level 5] Parsed ri git://github.com/mih/datalad into fields {'username': '', 'scheme': 'git', 'fragment': '', 'hostname': 'github.com', 'query': '', 'path': '/mih/datalad', 'password': '', 'port': ''}
[Level 5] Detected url ri
[Level 5] Parsed ri git://github.com/datalad/datalad into fields {'username': '', 'scheme': 'git', 'fragment': '', 'hostname': 'github.com', 'query': '', 'path': '/datalad/datalad', 'password': '', 'port': ''}
[Level 5] Detected url ri
[Level 8] Using existing Git repository at /home/yoh/proj/datalad/datalad-master
[Level 5] Parsed ri /home/yoh/proj/datalad/datalad-master into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/home/yoh/proj/datalad/datalad-master', 'password': '', 'port': ''}
[Level 5] Detected file ri
Exception IOError: IOError("Lock at '/home/yoh/proj/datalad/datalad-master/.git/index.lock' could not be obtained",) in <bound method AnnexRepo.__del__ of <AnnexRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.annexrepo.AnnexRepo'>)>> ignored
[Level 8] Using existing Git repository at /home/yoh/proj/datalad/datalad-master
Exception IOError: IOError("Lock at '/home/yoh/proj/datalad/datalad-master/.git/index.lock' could not be obtained",) in <bound method GitRepo.__del__ of <GitRepo path=/home/yoh/proj/datalad/datalad-master (<class 'datalad.support.gitrepo.GitRepo'>)>> ignored
[Level 5] Parsed ri /home/yoh/proj/datalad/datalad-master into fields {'username': '', 'scheme': '', 'fragment': '', 'hostname': '', 'query': '', 'path': '/home/yoh/proj/datalad/datalad-master', 'password': '', 'port': ''}
[Level 5] Detected file ri
[Level 9] Running: ['git', 'version']
```
I guess automagicio might need to acquire more of caching of previous results or smth like that...

**Comment by mih** [comment]
Looks as if this was addressed in #2050
